### ISSUE TITLE
Markdown headings don't render when prepending PowerShell release notes in wiki updater task

### ISSUE THREAD

**Issue by agilenut** [rfennell/AzurePipelines#826]
PowerShell Template:
```
## Version $($build.buildnumber)

@@CSLOOP@@
* $($csdetail.comment) ($($csdetail.changesetid)$($csdetail.commitid))
@@CSLOOP@@
```

Output:
![image](https://user-images.githubusercontent.com/705864/94832031-6b6da100-03db-11eb-9d81-b14b57cda103.png)

Where my cursor is, there seems to be a 0 width character between the new line and the first # sign. If I delete the character, I get this:
![image](https://user-images.githubusercontent.com/705864/94832121-82ac8e80-03db-11eb-82cf-1b99c8a3dce3.png)

I also noticed if I try copying the output into a file in Visual Studio and save, it says there are characters that can't be saved in Unicode.

I've already checked my template and re-typed it from scratch to ensure there was nothing from an odd copy-paste.

I can work around this by changing the template to this:
```
`n## Version $($build.buildnumber)

@@CSLOOP@@
* $($csdetail.comment) ($($csdetail.changesetid)$($csdetail.commitid))
@@CSLOOP@@
```

This works because it puts the odd character on a line of its own. Thus, the heading is on a new "clean" line and it renders correctly. 

I have not seen this issue with the Cross Platform version of the task but I currently cannot use the Cross Platform version in our on-prem instance because we are currently behind on the api version that it uses.

**Comment by rfennell** [comment]
I am thinking this could be down to file encoding. When the prepend is done it assumes UTF8 when reading the file contents

```
        var contents = fs.readFileSync(files[index], "utf8");
                if (appendToFile) {
                    fs.appendFileSync(targetFile, contents.replace(/`n/g, "\r\n"));
                    logInfo(`Appended to ${targetFile}`);
                } else {
                    var oldContent = "";
                    if (fs.existsSync(targetFile)) {
                        oldContent = fs.readFileSync(targetFile, "utf8");
                    }
                    fs.writeFileSync(targetFile, contents.replace(/`n/g, "\r\n"));
                    fs.appendFileSync(targetFile, oldContent);
                    logInfo(`Prepending to the ${targetFile}`);
                }

```

The question is what encoding are your file?

The quickest way to check open the file in notepad and do a 'save as' and the current encoding is shown.

If it this then the options are
1. Read the current encoding and keep using it
1. Allow the encoding to be passed as a parameter

**Comment by agilenut** [comment]
The file is written via the Generate Release Notes (PowerShell) step. We output to $(System.DefaultWorkingDirectory)\releasenotes.md. We don't do anything special to change the encoding. I'll see if I can get access to the file itself and verify it is UTF8. 

I did clone the wiki locally and made sure it is indeed UTF8.

**Comment by rfennell** [comment]
I found some time to look at this properly and have a good repo of the issue now. So I can sort out a fix

**Comment by rfennell** [comment]
The issue is not down to the source of the new content but the way Node seems to handle appending files. Even though both files involved in the append have no leading special character one appears to be added.

The fix I have come up with is to add a new parameter that selects whether this leading character is removed or not. To trim the character set `trimLeadingSpecialChar=true` (it is on the advanced panel if using a UI based build). 

**Note** I have defaulted this new setting to `false` so I don't introduce a breaking change for anyone.

This will be released as 1.17.x, it is currently going through my release pipeline

**Comment by rfennell** [comment]
Oh and 1.18.x fixes a potential issue parsing the retry count variable

**Comment by agilenut** [comment]
It looks like this new option just removes the first char of the existing wiki page. This does work correctly when the last version was produced by node. I did notice some of the files that I had edited, which did not have the leading character had the first "valid" character removed. In my case, it was a # sign for a markdown heading and removed the heading. 

Do you think there would be a way to detect if the leading character was the invalid one before stripping it?

**Comment by rfennell** [comment]
Yes a system that detects if it is special would be a nice improvement

I realise I have not seen the problem as I have blank line at the top and bottom if files

**Comment by rfennell** [comment]
I have altered to the code to check that the character to be delete is the 65279 Unicode Character 'ZERO WIDTH NO-BREAK SPACE' special character. This is the only one deleted if the new parameter is set.

This is shipped as 1.19.x
